File: pt5/1188.design-bounded-blocking-queue.go
package main

import (
	"container/list"
	"sync"
)

type BoundedBlockingQueue struct {
	slots chan struct{}
	items chan struct{}
	mu    sync.Mutex
	data  *list.List
}

func Constructor(capacity int) BoundedBlockingQueue {
	slots := make(chan struct{}, capacity)
	for i := 0; i < capacity; i++ {
		slots <- struct{}{}
	}
	return BoundedBlockingQueue{
		slots: slots,
		items: make(chan struct{}, capacity),
		data:  list.New(),
	}
}

func (q *BoundedBlockingQueue) Enqueue(element int) {
	<-q.slots
	q.mu.Lock()
	q.data.PushFront(element)
	q.mu.Unlock()
	q.items <- struct{}{}
}

func (q *BoundedBlockingQueue) Dequeue() int {
	<-q.items
	q.mu.Lock()
	back := q.data.Back()
	value := back.Value.(int)
	q.data.Remove(back)
	q.mu.Unlock()
	q.slots <- struct{}{}
	return value
}

func (q *BoundedBlockingQueue) Size() int {
	q.mu.Lock()
	defer q.mu.Unlock()
	return q.data.Len()
}

File: pt5/1188.design-bounded-blocking-queue.java
import java.util.*;
import java.util.concurrent.Semaphore;
import java.util.concurrent.locks.ReentrantLock;

class BoundedBlockingQueue {
    private final int capacity;
    private final Deque<Integer> queue = new ArrayDeque<>();
    private final Semaphore emptySlots;
    private final Semaphore availableItems = new Semaphore(0);
    private final ReentrantLock lock = new ReentrantLock();

    public BoundedBlockingQueue(int capacity) {
        this.capacity = capacity;
        this.emptySlots = new Semaphore(capacity);
    }

    public void enqueue(int element) throws InterruptedException {
        emptySlots.acquire();
        lock.lock();
        try {
            queue.addFirst(element);
        } finally {
            lock.unlock();
        }
        availableItems.release();
    }

    public int dequeue() throws InterruptedException {
        availableItems.acquire();
        lock.lock();
        try {
            return queue.removeLast();
        } finally {
            lock.unlock();
            emptySlots.release();
        }
    }

    public int size() {
        lock.lock();
        try {
            return queue.size();
        } finally {
            lock.unlock();
        }
    }
}

